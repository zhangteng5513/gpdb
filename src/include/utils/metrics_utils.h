/*-------------------------------------------------------------------------
 *
 * metrics_utils.h
 *	  Definitions for query metrics struct and functions
 *
 * Portions Copyright (c) 2017-Present Pivotal Software, Inc.
 *
 *
 * IDENTIFICATION
 *	    src/include/utils/metrics_utils.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef METRICS_UTILS_H
#define METRICS_UTILS_H 

typedef enum MetricsStatus
{	
	METRICS_NODE_INITIALIZE = 100,
	METRICS_NODE_EXECUTING,
	METRICS_NODE_FINISHED,
	
	METRICS_QUERY_SUBMIT = 200,
	METRICS_QUERY_START,
	METRICS_QUERY_DONE,
	METRICS_QUERY_ERROR,
	METRICS_QUERY_CANCELING,
	METRICS_QUERY_CANCELED
} MetricsStatus;

typedef void (*query_metrics_entry_hook_type)(MetricsStatus, void *);
extern query_metrics_entry_hook_type query_metrics_entry_hook;

#endif   /* METRICS_UTILS_H */
