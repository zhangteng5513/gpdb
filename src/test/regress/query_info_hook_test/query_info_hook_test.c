#include "postgres.h"

#include "fmgr.h"
#include "utils/metrics_utils.h"

PG_MODULE_MAGIC;

void _PG_init(void);
void _PG_fini(void);

static query_info_collect_hook_type prev_query_info_collect_hook;

static void
test_hook(QueryMetricsStatus, void* args);

void
_PG_init(void)
{
	prev_query_info_collect_hook = query_info_collect_hook;
	query_info_collect_hook = test_hook;
}

void
_PG_fini(void)
{
		query_info_collect_hook = prev_query_info_collect_hook;
}

static void
test_hook(QueryMetricsStatus status, void* args)
{
	switch (status)
	{
		case METRICS_PLAN_NODE_INITIALIZE:
			ereport(LOG, (errmsg("Plan node initializing")));
			break;
		case METRICS_PLAN_NODE_EXECUTING:
			ereport(LOG, (errmsg("Plan node executing")));
			break;
		case METRICS_PLAN_NODE_FINISHED:
			ereport(LOG, (errmsg("Plan node finished")));
			break;
		case METRICS_QUERY_SUBMIT:
			ereport(LOG, (errmsg("Query submit")));
			break;
		case METRICS_QUERY_START:
			ereport(LOG, (errmsg("Query start")));
			break;
		case METRICS_QUERY_DONE:
			ereport(LOG, (errmsg("Query Done")));
			break;
		case METRICS_QUERY_ERROR:
			ereport(LOG, (errmsg("Query Error")));
			break;
		case METRICS_QUERY_CANCELING:
			ereport(LOG, (errmsg("Query Canceling")));
			break;
		case METRICS_QUERY_CANCELED:
			ereport(LOG, (errmsg("Query Canceled")));
			break;
	}
}
