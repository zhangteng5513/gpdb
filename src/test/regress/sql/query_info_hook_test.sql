LOAD '/home/gpadmin/workspace_wh/gpdb_metrics/src/test/regress/query_info_hook_test/query_info_hook_test.so';
SET client_min_messages='log';

-- Test normal case
SELECT 1 AS a;

--Test Error case
SELECT 1/0 AS a;
