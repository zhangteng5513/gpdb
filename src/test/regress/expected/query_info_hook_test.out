LOAD '/home/gpadmin/workspace_wh/gpdb_metrics/src/test/regress/query_info_hook_test/query_info_hook_test.so';
SET client_min_messages='log';
-- Test normal case
SELECT 1 AS a;
LOG:  statement: SELECT 1 AS a;
LOG:  Query submit
LOG:  Query start
LOG:  Plan node initializing
LOG:  Plan node executing
LOG:  Plan node finished
LOG:  Query Done
 a 
---
 1
(1 row)

--Test Error case
SELECT 1/0 AS a;
LOG:  statement: SELECT 1/0 AS a;
LOG:  Query submit
LOG:  Query start
LOG:  Plan node initializing
LOG:  Plan node executing
LOG:  Query Error
LOG:  Query Error
ERROR:  division by zero
